### PR TITLE
hacky method to exclude items that never freeze or rot from the texture atlas

### PR DESCRIPTION
**Pull request by Mervius**
A potential fix for https://github.com/Viola-Siemens/Freeze-It-And-Heat-It/issues/30
This is kind of horrible right now, but is an example and functional implementation of not adding items in the never_freeze or never_rotten blacklists to the texture atlas. This absolutely could be written better.

This also has to test if an item name is "" because minecraft apparently allows empty file names as resources...
So if a mod has an item resource that's an empty file name, right now, it will always be accepted because I don't know a good way to add that to the config and there's a chance it might be food.

The code is also a bit ugly so that the never_frozen and never_rotten configs that come with the mod can still use the same "modid:item" in this detection, there might be a cleaner way to do it.

This is a draft because there are probably better ways to do this.

### CODE REVIEW

**Comment by Mervius** [comment]
This still relies on adding the problematic files to the config, but is better than nothing.

**Comment by Mervius** [comment]
By adding the problematic 4096x4096 item to both "NEVER_ROTTEN_FOODS" and "NEVER_FROZEN_FOODS", my texture atlas turns from this:
489196479-9552071f-d591-44c2-a15e-01bed06b21f1
to this:
minecraft_textures_atlas_blocks png_3
Which, if done for every item that isn't food or is unreasonably large, would significantly reduce the space being taken up in the atlas by junk

**Comment by Mervius** [comment]
One issue with this implementation is that the config list would grow very large if you added every non-food item to it, which *(might?)* cause memory issues. A potential way to fix that is blacklisting an entire mod instead, then manually whitelisting specific items from it if needed, as that should generally take up less array elements than what this pull request does. That would need more config options, though.

Adding a lot of items to the exclusion config is probably unnecessary, since you could just add a few problem items until your modpack stops crashing. Evidence of this is the image above where I added a single item to the config and regained over 300-thousand 16x16 pixel areas for other items.

**Comment by Mervius** [comment]
The quadruple nested if statements are also something I want to try to fix and clean up, but with there being "item/" and "items/" as options and empty file names needing to be accounted for, that was what I came up with for now.

**Comment by Mervius** [comment]
I made the if statements better by changing some checks to using regex. which I tried to do initially but it failed... this time it worked...

**Comment by Mervius** [comment]
Also added some logging so that people can know which items are large and potentially causing issues

**Comment by Mervius** [comment]
Note that the "namespace:item" that you put into the config and is output in the logs isn't necessarily going to be an actual proper "item" that gets registered in-game, it's just the resource name(without the "item/" or "items/"), so the other parts of this mod that interact with the NEVER_ROTTEN_FOODS or NEVER_FROZEN_FOODS might behave weirdly with them if you aren't careful with how you use them, but the current implementation doesn't *seem* to have any problems(outside of the CME but that is unrelated to this issue.)

However, it does mean that if an item you want to exclude has a resource name different from it's in-game registered name/namespace, it could potentially have to be in the config twice. I do not know how common that is.

File: src/main/java/com/hexagram2021/fiahi/mixin/SpriteLoaderMixin.java [hunk 1]
```diff
@@ -47,7 +47,7 @@ public abstract class SpriteLoaderMixin {
 								skip_frozen = true;
 							} if (FIAHICommonConfig.NEVER_ROTTEN_FOODS.get().contains(content.name().getNamespace() + ":" + content.name().getPath().split("items?\\/")[1])) {
 								skip_rotten = true;
-							} if ((content.width() > 16 || content.height() > 16) && !skip_frozen %% !skip_rotten) {
```
Comment (Mervius) [inline src/main/java/com/hexagram2021/fiahi/mixin/SpriteLoaderMixin.java]: Idk how this happened, tbh

**Comment by Mervius** [comment]
This implementation does technically cause missing textures if you manually set an item on the blacklist to have a temperature value it's blacklisted from, but that *shouldn't* happen in normal gameplay


EDIT: It can happen if you have a frozen or rotten item then add that item to the blacklists in the config after the fact. I'm not sure if that qualifies as "normal gameplay", though

**Comment by Mervius** [comment]
I "fixed" the missing textures for items forced to be frozen or rotten with commands when they are excluded from freezing or rotting by just using the normal texture for those items.

Ideally, I think it would be better to just force the frozen or rotten level back to 0 for these items, but I have not figured that out yet

**Comment by Mervius** [comment]
OK, i think I fixed the missing textures properly, now they aren't even created to show up in the log as missing.

Before I just wasn't using them, now I am using them, but they all map to the same normal texture

**Comment by Mervius** [comment]
If anyone knows a better way to get a namespace and path as a string during stitching and model baking than `resourceLocation.getNamespace() + ":" + resourceLocation.name().getPath()` I'd love to see it.

We only have access to the texture atlas and model names, which means you can't use the "normal" way to get them by using an item class

**Comment by Mervius** [comment]
If there isn't a better way, then I think this might be ready for review, but I don't want to do that if there is a better way

File: src/main/java/com/hexagram2021/fiahi/mixin/ModelBakeryMixin.java [hunk 2]
```diff
@@ -41,15 +42,24 @@ public abstract class ModelBakeryMixin {
...
 			if (unbakedModel instanceof BlockModel && ((BlockModel) unbakedModel).getRootModel() == ModelBakery.GENERATION_MARKER) {
 				this.fiahi$putBakedModel(
 						spriteId,
-						this.fiahi$bakeModel(spriteMapper, ".frozen.1", spriteId),
-						this.fiahi$bakeModel(spriteMapper, ".frozen.2", spriteId),
-						this.fiahi$bakeModel(spriteMapper, ".frozen.3", spriteId),
-						this.fiahi$bakeModel(spriteMapper, ".rotten.1", spriteId),
-						this.fiahi$bakeModel(spriteMapper, ".rotten.2", spriteId),
-						this.fiahi$bakeModel(spriteMapper, ".rotten.3", spriteId)
+						this.fiahi$bakeModel(spriteMapper, String.format(skip_frozen ? "" : ".frozen.1"), spriteId),
```
Comment (Mervius) [inline src/main/java/com/hexagram2021/fiahi/mixin/ModelBakeryMixin.java]: Just realized I didn't do what I actually was trying to do here... oops. It still works, just not quite what I was trying to do.

**Comment by Mervius** [comment]
I added regex support to the config so that you can also remove items with multiple textures like the clock or compass without having to specify each version of the texture manually. This feature can crash the game or cause strange things if the person writing the regex isn't completely certain of what they are doing.

It does not currently prevent edible items from freezing or rotting if they match the regex, just from changing sprites. Adding regex support to the freezing and rotting in-game would be a bit more work.

For example: "(minecraft:.+)" will match against literally every item in vanilla, and prevent that frozen/rotten sprite from existing or being used, but the matched food will freeze or rot even though it doesn't have sprites made for that...

I am working on resolving that

**Comment by Mervius** [comment]
Regex should function "properly" now, there may still be some bugs... and the game crashes if you don't escape your string properly when using regex...

File: src/main/java/com/hexagram2021/fiahi/common/item/capability/IFrozenRottenFood.java [hunk 3]
```diff
@@ -49,11 +49,22 @@ default void apply(double newTemperature, @Nullable Item item) {
...
+					}
+				}
+				for (String entry : FIAHICommonConfig.NEVER_ROTTEN_FOODS.get()) {
+					if (item_name.matches(entry)) {
+						skip_rotten = true;
+					}
+				}
+				if(flag && item != null && skip_rotten) {
 					this.setTemperature(FROZEN_ROTTEN_THRESHOLD * 2 - EPS);
-				} else if(!flag && item != null &&
```
Comment (Mervius) [inline src/main/java/com/hexagram2021/fiahi/common/item/capability/IFrozenRottenFood.java:55]: Why is this an else if?

**Comment by Mervius** [comment]
Oh ok... renaming the branch doesn't work properly and update the reference here... kind of dumb...

**Comment by Mervius** [comment]
These changes would still need to be made for the other branches, and with the way this repository is set up, would be a bit more work than just making a pr from this repo to the others.
